### PR TITLE
[SYCL] Implement device library linking for runtime compilation

### PR DESCRIPTION
**Pull request by jopperm**
Mimics `clang::driver::tools::SYCL::getDeviceLibraries` assuming a SPIR-V target (= no AoT, no third-party GPUs, no native CPU).

Same as for the compilation step, warning/error reporting is still rudimentary and will be improved in a future PR.

### CODE REVIEW

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 1]
```diff
@@ -96,23 +103,20 @@ struct GetLLVMModuleAction : public ToolAction {
...
   const std::string &DPCPPRoot = getDPCPPRoot();
   if (DPCPPRoot == InvalidDPCPPRoot) {
-    return llvm::createStringError("Could not locate DPCPP root directory");
+    return createStringError("Could not locate DPCPP root directory");
   }
 
   SmallVector<std::string> CommandLine = {"-fsycl-device-only"};
-  // TODO: Allow instrumentation again when device library linking is
-  //       implemented.
-  CommandLine.push_back("-fno-sycl-instrument-device-code");
```
Comment (jopperm) [inline sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:111]: NB: This PR is tested by the existing E2E, which now works with device instrumentation enabled.

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 2]
```diff
@@ -143,5 +147,90 @@ jit_compiler::compileDeviceCode(InMemoryFile SourceFile,
...
+      continue;
+    }
+    if (UA == "-fsycl-instrument-device-code") {
+      DeviceInstrumentationEnabled = true;
+      continue;
+    }
+
+    // Issue warning for `-fsycl-device-lib` or `-fno-sycl-device-lib`.
+    // TODO: Is it worth supporting these flags? We're using `LinkOnlyNeeded`
+    //       mode anyways!
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:173]: Do we currently link against the libraries supported by this flag (`libm-fp32, libm-fp64, libc`)?

If not, should we support the flag to allow users to link them?

If we link, should we support the negative version of the flag (`-fno-sycl-device-lib`) to reduce overhead?
The overhead is probably not super high, but finding the file on disk, loading it into memory and then linking is also not free.

**Review by sommerlukas — APPROVED** [review]

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jopperm) [reply sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:173]: > Do we currently link against the libraries supported by this flag (`libm-fp32, libm-fp64, libc`)?

Yes, the default is to link all libraries.

> If we link, should we support the negative version of the flag (-fno-sycl-device-lib) to reduce overhead?

Yeah, good point. Then I think I'll just go ahead and implement the handling for these flags. Makes the need to emit a warning here go away, and I'll have to inspect user args again for the post-link transformations anyways.

**Comment by jopperm** [comment]
Reverted to draft while working on better handling of user arguments.

**Comment by jopperm** [comment]
@cperkinsintel @gmlueck @AlexeySachkov could you take a look, please?

**Review by cperkinsintel — APPROVED** [review]

**Comment by gmlueck** [comment]
> @cperkinsintel @gmlueck @AlexeySachkov could you take a look, please?

Sorry, I can't commit to reviewing every PR for the jit compiler.  I generally review changes to specifications, and I'm available to answer questions.

**Pull request merged**